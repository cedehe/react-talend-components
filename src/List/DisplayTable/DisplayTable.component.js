import React, { PropTypes } from 'react';
import classnames from 'classnames';
import { Button } from 'react-bootstrap';
import Actions from '../../Actions';
import Icon from '../../Icon';

import theme from './DisplayTable.scss';

function RowRenderer(props) {
	const { titleKey, onTitleClick, item } = props;
	return (
		<tr>
			{props.columns.map((column, index) => {
				if (column.key === titleKey) {
					const onClick = event => onTitleClick(
						event,
						item,
					);
					const iconName = props.iconKey && props.item[props.iconKey];
					return (
						<td key={index}>
							{iconName && <Icon name={iconName} />}
							<Button
								bsStyle="link"
								onClick={onClick}
								role="link"
							>
								{item[column.key]}
							</Button>
							<Actions
								actions={item.actions || []}
								hideLabel
								link
							/>
						</td>
					);
				}
				return (<td key={index}>{item[column.key]}</td>);
			})}
		</tr>
	);
}
RowRenderer.propTypes = {
	item: PropTypes.object, // eslint-disable-line react/forbid-prop-types
	columns: PropTypes.arrayOf(
		PropTypes.shape({
			key: PropTypes.string.isRequired,
		})
	).isRequired,
	iconKey: PropTypes.string,
	titleKey: PropTypes.string,
	onTitleClick: PropTypes.func,
};

function ListHeader(props) {
	return (
		<tr>
			{props.columns.map((column, index) => (<th key={index}>{column.label}</th>))}
		</tr>
	);
}
ListHeader.propTypes = {
	columns: PropTypes.arrayOf(
		PropTypes.shape(
			{ label: PropTypes.string },
		)
	),
};


/**
 * @param {object} props react props
 * @example
 <DisplayTable name="Hello world"></DisplayTable>
 */
function DisplayTable({ items, columns, iconKey, titleKey, onTitleClick }) {
	const className = classnames(
		'table',
		'tc-list-display-table',
		theme.table,
	);
	return (
		<table className={className}>
			<thead>
				<ListHeader columns={columns} />
			</thead>
			<tbody>
				{items.map(
					(item, index) => (
						<RowRenderer
							key={index}
							item={item}
							columns={columns}
							iconKey={iconKey}
							titleKey={titleKey}
							onTitleClick={onTitleClick}
						/>
					)
				)}
			</tbody>
		</table>
	);
}

DisplayTable.propTypes = {
	items: PropTypes.arrayOf(
		PropTypes.object
	),
	columns: PropTypes.arrayOf(
		PropTypes.object
	),
	iconKey: PropTypes.string,
	titleKey: PropTypes.string,
	onTitleClick: PropTypes.func,
};

DisplayTable.defaultProps = {
	items: [],
	titleKey: 'name',
};

export default DisplayTable;
