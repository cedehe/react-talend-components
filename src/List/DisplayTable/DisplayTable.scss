// DisplayTable styles
$tc-list-table-hover-bg: $concrete !default;
$tc-list-table-color: $dove-gray !default;
$tc-list-table-hover-color: $black !default;

.table :global(.btn-group) {
	visibility: hidden;
}

.table tr:hover :global(.btn-group) {
	visibility: visible;
}

.table :global(.btn-link) {
	color: $tc-list-table-color;
	font-size: inherit;
	letter-spacing: inherit;
	font-weight: 600;
}

.table :global(thead tr) {
	text-transform: uppercase;
}

.table :global(tbody tr:hover) {
	color: $tc-list-table-hover-color;
	background-color: $tc-list-table-hover-bg;
}

.table :global(.btn-link:hover) {
	color: $tc-list-table-hover-color;
	text-decoration: none;
}

