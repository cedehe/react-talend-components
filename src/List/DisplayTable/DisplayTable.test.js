import React from 'react';
import renderer from 'react-test-renderer';

import DisplayTable from './DisplayTable.component';

jest.mock('react-dom');
jest.mock('uuid');

const items = [
	{
		id: 1,
		name: 'Hello world',
		created: '2016-09-22',
		modified: '2016-09-22',
		author: 'Jean-Pierre DUPONT',
		actions: [{
			label: 'edit',
			icon: 'fa fa-edit',
			onClick: jest.fn(),
		}, {
			label: 'delete',
			icon: 'fa fa-trash-o',
			onClick: jest.fn(),
		}],
		icon: 'fa fa-file-excel-o',
	},
	{
		id: 2,
		name: 'Foo',
		created: '2016-09-22',
		modified: '2016-09-22',
		author: 'Jean-Pierre DUPONT',
		icon: 'fa fa-file-pdf-o',
	},
	{
		id: 2,
		name: 'Bar',
		created: '2016-09-22',
		modified: '2016-09-22',
		author: 'Jean-Pierre DUPONT',
	},
];

const columns = [
	{ key: 'id', label: 'Id' },
	{ key: 'name', label: 'Name' },
	{ key: 'author', label: 'Author' },
	{ key: 'created', label: 'Created' },
	{ key: 'modified', label: 'Modified' },
];

describe('DisplayTable', () => {
	it('should render with defined title property', () => {
		// given
		const props = {
			items,
			columns,
			titleKey: 'id', // title key defined
			iconKey: 'icon',
			onTitleClick: jest.fn(),
		};

		// when
		const wrapper = renderer.create(<DisplayTable {...props} />).toJSON();

		// then
		expect(wrapper).toMatchSnapshot();
	});

	it('should render with default title property (name)', () => {
		// given
		const props = {
			items,
			columns,
			titleKey: undefined, // no title key defined
			iconKey: 'icon',
			onTitleClick: jest.fn(),
		};

		// when
		const wrapper = renderer.create(<DisplayTable {...props} />).toJSON();

		// then
		expect(wrapper).toMatchSnapshot();
	});
});
