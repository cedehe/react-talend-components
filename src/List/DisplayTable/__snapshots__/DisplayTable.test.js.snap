exports[`DisplayTable should render with default title property (name) 1`] = `
<table
  className="table tc-list-display-table">
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Name
      </th>
      <th>
        Author
      </th>
      <th>
        Created
      </th>
      <th>
        Modified
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        <i
          aria-hidden="true"
          className="fa fa-file-excel-o"
          title={null} />
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          role="link"
          type="button">
          Hello world
        </button>
        <div
          className="tc-actions btn-group">
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              className="fa fa-edit" />
          </button>
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              className="fa fa-trash-o" />
          </button>
        </div>
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        <i
          aria-hidden="true"
          className="fa fa-file-pdf-o"
          title={null} />
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          role="link"
          type="button">
          Foo
        </button>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          role="link"
          type="button">
          Bar
        </button>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`DisplayTable should render with defined title property 1`] = `
<table
  className="table tc-list-display-table">
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Name
      </th>
      <th>
        Author
      </th>
      <th>
        Created
      </th>
      <th>
        Modified
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <i
          aria-hidden="true"
          className="fa fa-file-excel-o"
          title={null} />
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          role="link"
          type="button">
          1
        </button>
        <div
          className="tc-actions btn-group">
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              className="fa fa-edit" />
          </button>
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              className="fa fa-trash-o" />
          </button>
        </div>
      </td>
      <td>
        Hello world
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        <i
          aria-hidden="true"
          className="fa fa-file-pdf-o"
          title={null} />
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          role="link"
          type="button">
          2
        </button>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Foo
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          role="link"
          type="button">
          2
        </button>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Bar
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
  </tbody>
</table>
`;
